### PR TITLE
Add support to force user specifying reason for requests which trigger operation.

### PR DESCRIPTION
**Pull request by kidkun**
Fix issue #977

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaCruiseControlConfig.java [hunk 1]
```diff
@@ -934,6 +935,13 @@
   private static final String TWO_STEP_PURGATORY_MAX_REQUESTS_DOC = "The maximum number of requests in two-step "
       + "(verification) purgatory.";
 
+  /**
+   * <code>request.reason.required</code>
+   */
+  public static final String REQUEST_REASON_REQUIRED_CONFIG = "request.reason.required";
+  private static final String REQUEST_REASON_REQUIRED_DOC = "Require specifying reason via " + REASON_PARAM + " parameter "
+      + "for non-dryrun POST request";
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaCruiseControlConfig.java]: Is this config controls this behavior for _all_  or _selected_ POST requests -- i.e. should `stop_proposal_execution` get a `reason` parameter, as well? If it is only for selected endpoints, can we enumerate them?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java [hunk 2]
```diff
@@ -93,7 +93,8 @@ public void configure(Map<String, ?> configs) {
                                                            allowCapacityEstimation,
                                                            excludeRecentlyDemotedBrokers,
                                                            excludeRecentlyRemovedBrokers,
-                                                           _anomalyId)
+                                                           _anomalyId,
+                                                           String.format("Self healing for broker failure : %s", this))
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java]: Nit (applies to similar uses in other anomaly types for the sake of consistency): `broker failure` -> `%s` that uses `AnomalyType.BROKER_FAILURE`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java [hunk 3]
```diff
@@ -20,13 +19,14 @@
...
 import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.getSelfHealingGoalNames;
 import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.KAFKA_CRUISE_CONTROL_OBJECT_CONFIG;
+import static com.linkedin.kafka.cruisecontrol.detector.notifier.AnomalyType.BROKER_FAILURE;
 
 
 /**
  * The broker failures that have been detected.
  */
 public class BrokerFailures extends KafkaAnomaly implements CruiseControlConfigurable {
-  protected static final String ID_PREFIX = AnomalyType.BROKER_FAILURE.toString();
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java]: Now that we have a separation between the following:
```
  TRIGGERED_USER_TASK_ID = "triggeredUserTaskId";
  TRIGGERED_SELF_HEALING_TASK_ID = "triggeredSelfHealingTaskId";
  TRIGGERED_TASK_REASON = "triggeredTaskReason";
```

I am curious if we can drop `ID_PREFIX` overwrite from individual anomaly types (and handle the drop accordingly in the rest of the codebase) -> thus have an `anomalyId` that is a proper `uuid`?

Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java]: we have some places in our code depend on the assumption that self-healing task's uuid starts with anomaly type. 

One example is [determine whether mark `AnomalyDetector` or `UserTaskManager` when execution starts](https://github.com/linkedin/cruise-control/blob/d10aa73209ce0f542afca44ad7145e7b8fe8112d/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java#L699).
If we remove the prefix, we need to make decision based on `_reason` field, we can chose a reason template for self-healing tasks; but for user-triggered task, the `_reason` is provided by user, so theoretically user can accidentally provide a reason which matchs our pre-defined self-healing reason template.

Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java]: Why not change the places that depends on `uuid` prefix? For the same reason, I don't think we should rely on parsing the `_reason` field.

For the specific example above, can't we just pass the information regarding `who` started the execution? Internally, we always know whether the execution was triggered by self healing or not; hence, it is possible to pass this information down to the executor.

Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java]: That's a good idea! 
I removed the prefix  in UUID and pass down the information of who triggers operation from `Runnable` to `Executor`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 4]
```diff
@@ -692,16 +698,14 @@ public void run() {
...
       // retrieve the associated user task information.
-      if (AnomalyType.cachedValues().stream().anyMatch(type -> _uuid.startsWith(type.toString()))) {
-        userTaskInfo = null;
-      } else {
+      if (_isTriggeredByUserRequest) {
         userTaskInfo = _userTaskManager.markTaskExecutionBegan(_uuid);
       }
       _state = STARTING_EXECUTION;
-      _executorState = ExecutorState.executionStarted(_uuid, _reason, _recentlyDemotedBrokers, _recentlyRemovedBrokers);
+      _executorState = ExecutorState.executionStarted(_uuid, _reason, _recentlyDemotedBrokers, _recentlyRemovedBrokers, _isTriggeredByUserRequest);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Can we also log the `_reason` in the  `OPERATION_LOG` (please see the next line)?

**Pull request merged**